File: runtime/java/src/main/java/dev/argon/verilization/runtime/Option.java
package dev.argon.verilization.runtime;

import java.util.Optional;
import java.io.IOException;

public final class Option {
    private Option() {}

    public static <A> Optional<A> fromCaseSome(A value) {
        return Optional.of(value);
    }

    public static <A> Optional<A> fromCaseNone() {
        return Optional.empty();
    }

    public static <A, B> Converter<Optional<A>, Optional<B>> converter(Converter<A, B> elementConverter) {
        return prev -> prev.map(elementConverter::convert);
    }

    public static <T> Codec<Optional<T>> codec(Codec<T> elementCodec) {
        return new Codec<Optional<T>>() {
            @Override
            public Optional<T> read(FormatReader reader) throws IOException {
                boolean present = reader.readByte() != 0;
                if(present) {
                    return Optional.of(elementCodec.read(reader));
                }
                else {
                    return Optional.empty();
                }
            }
    
            @Override
            public void write(FormatWriter writer, Optional<T> value) throws IOException {
                if(value.isPresent()) {
                    writer.writeByte((byte)1);
                    elementCodec.write(writer, value.get());
                }
                else {
                    writer.writeByte((byte)0);
                }
            }
    
        };
    }
}

File: runtime/typescript/src/Option.ts
import { Converter } from "./Converter.js";
import { FormatReader, FormatWriter } from "./FormatIO.js";
import { Codec } from "./Codec.js";

export type Option<A> = { readonly value: A; } | null;

export function fromCaseSome<A>(value: A): Option<A> {
    return { value };
}

export function fromCaseNone<A>(): Option<A> {
    return null;
}

export function converter<A, B>(elementConverter: Converter<A, B>): Converter<Option<A>, Option<B>> {
    return {
        convert(prev: Option<A>): Option<B> {
            if(prev === null) {
                return null;
            }
            else {
                return { value: elementConverter.convert(prev.value) };
            }
        }
    };
}

export function codec<A>(elementCodec: Codec<A>): Codec<Option<A>> {
    return {
        async read(reader: FormatReader): Promise<Option<A>> {
            const present = await reader.readU8();

            if(present !== 0) {
                const value = await elementCodec.read(reader);
                return { value };
            }
            else {
                return null;
            }
        },

        async write(writer: FormatWriter, value: Option<A>): Promise<void> {
            if(value !== null) {
                await writer.writeU8(1);
                await elementCodec.write(writer, value.value);
            }
            else {
                await writer.writeU8(0);
            }
        },
    };
}
